File: frontend/src/lib/api/schedules.ts
/**
 * Scheduled Builds API types, endpoint functions, and React Query hooks.
 *
 * Backend route reference:
 *   GET    /api/v1/schedules                    — list schedules
 *   GET    /api/v1/schedules/:id                — schedule detail with executions
 *   POST   /api/v1/schedules                    — create schedule
 *   PUT    /api/v1/schedules/:id                — update schedule
 *   DELETE /api/v1/schedules/:id                — delete schedule
 *   POST   /api/v1/schedules/:id/pause          — pause schedule
 *   POST   /api/v1/schedules/:id/resume         — resume schedule
 *   POST   /api/v1/schedules/:id/run-now        — trigger immediate execution
 *   GET    /api/v1/schedules/:id/executions     — execution history
 */

import {
  useMutation,
  useQuery,
  useQueryClient,
} from '@tanstack/react-query';

import apiClient from './client';

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

/** Schedule status values. */
export type ScheduleStatus = 'active' | 'paused' | 'completed' | 'expired';

/** The type of schedule: one-time or recurring. */
export type ScheduleType = 'one_time' | 'recurring';

/** Execution status values. */
export type ExecutionStatus =
  | 'pending'
  | 'running'
  | 'completed'
  | 'failed'
  | 'skipped';

/** Server configuration embedded in a schedule. */
export interface ScheduleServerConfig {
  environment?: string;
  os_type?: string;
  instance_size?: string;
  region?: string;
  puppet_role?: string;
  template_id?: string;
  [key: string]: unknown;
}

/** A build schedule returned by the API. */
export interface BuildSchedule {
  id: string;
  name: string;
  description: string | null;
  status: ScheduleStatus;
  schedule_type: ScheduleType;
  scheduled_at: string | null;
  cron_expression: string | null;
  timezone: string;
  next_run_at: string | null;
  last_run_at: string | null;
  run_count: number;
  max_runs: number | null;
  server_config: ScheduleServerConfig | null;
  created_by: string | null;
  created_at: string;
  updated_at: string;
}

/** A single execution record for a schedule. */
export interface ScheduleExecution {
  id: string;
  schedule_id: string;
  server_request_id: string | null;
  status: ExecutionStatus;
  scheduled_for: string;
  started_at: string | null;
  completed_at: string | null;
  error_message: string | null;
}

/** Response from the list schedules endpoint. */
export interface ScheduleListResponse {
  schedules: BuildSchedule[];
  total: number;
}

/** Query parameters for listing schedules. */
export interface ScheduleListParams {
  status?: ScheduleStatus;
  schedule_type?: ScheduleType;
  skip?: number;
  limit?: number;
}

/** Request body for creating a schedule. */
export interface ScheduleCreateRequest {
  name: string;
  description?: string;
  schedule_type: ScheduleType;
  scheduled_at?: string;
  cron_expression?: string;
  timezone: string;
  max_runs?: number;
  server_config: ScheduleServerConfig;
}

/** Request body for updating a schedule. */
export interface ScheduleUpdateRequest {
  name?: string;
  description?: string;
  scheduled_at?: string;
  cron_expression?: string;
  timezone?: string;
  max_runs?: number;
  server_config?: ScheduleServerConfig;
}

/** Execution list response. */
export interface ExecutionListResponse {
  executions: ScheduleExecution[];
  total: number;
}

// ---------------------------------------------------------------------------
// Query key factory
// ---------------------------------------------------------------------------

export const scheduleKeys = {
  all: ['schedules'] as const,
  list: (params?: ScheduleListParams) => ['schedules', 'list', params] as const,
  detail: (id: string) => ['schedules', id] as const,
  executions: (id: string) => ['schedules', id, 'executions'] as const,
} as const;

// ---------------------------------------------------------------------------
// Endpoint functions
// ---------------------------------------------------------------------------

/** List schedules with optional filtering and pagination. */
export async function listSchedules(
  params?: ScheduleListParams,
): Promise<ScheduleListResponse> {
  const response = await apiClient.get<ScheduleListResponse>(
    '/schedules',
    { params },
  );
  return response.data;
}

/** Get a single schedule by ID. */
export async function getSchedule(id: string): Promise<BuildSchedule> {
  const response = await apiClient.get<BuildSchedule>(`/schedules/${id}`);
  return response.data;
}

/** Create a new build schedule. */
export async function createSchedule(
  data: ScheduleCreateRequest,
): Promise<BuildSchedule> {
  const response = await apiClient.post<BuildSchedule>('/schedules', data);
  return response.data;
}

/** Update an existing build schedule. */
export async function updateSchedule(
  id: string,
  data: ScheduleUpdateRequest,
): Promise<BuildSchedule> {
  const response = await apiClient.put<BuildSchedule>(
    `/schedules/${id}`,
    data,
  );
  return response.data;
}

/** Delete a schedule. */
export async function deleteSchedule(id: string): Promise<void> {
  await apiClient.delete(`/schedules/${id}`);
}

/** Pause an active schedule. */
export async function pauseSchedule(id: string): Promise<BuildSchedule> {
  const response = await apiClient.post<BuildSchedule>(
    `/schedules/${id}/pause`,
  );
  return response.data;
}

/** Resume a paused schedule. */
export async function resumeSchedule(id: string): Promise<BuildSchedule> {
  const response = await apiClient.post<BuildSchedule>(
    `/schedules/${id}/resume`,
  );
  return response.data;
}

/** Trigger an immediate execution of a schedule. */
export async function runScheduleNow(id: string): Promise<ScheduleExecution> {
  const response = await apiClient.post<ScheduleExecution>(
    `/schedules/${id}/run-now`,
  );
  return response.data;
}

/** List execution history for a schedule. */
export async function listScheduleExecutions(
  id: string,
): Promise<ExecutionListResponse> {
  const response = await apiClient.get<ExecutionListResponse>(
    `/schedules/${id}/executions`,
  );
  return response.data;
}

// ---------------------------------------------------------------------------
// React Query hooks
// ---------------------------------------------------------------------------

/**
 * Fetch paginated schedule list.
 * Polls every 30 seconds for schedules with upcoming runs.
 */
export function useSchedules(params?: ScheduleListParams) {
  return useQuery<ScheduleListResponse>({
    queryKey: scheduleKeys.list(params),
    queryFn: () => listSchedules(params),
    refetchInterval: 30_000,
  });
}

/** Fetch a single schedule by ID. */
export function useSchedule(id: string | undefined) {
  return useQuery<BuildSchedule>({
    queryKey: scheduleKeys.detail(id!),
    queryFn: () => getSchedule(id!),
    enabled: !!id,
  });
}

/** Fetch execution history for a schedule. */
export function useScheduleExecutions(id: string | undefined) {
  return useQuery<ExecutionListResponse>({
    queryKey: scheduleKeys.executions(id!),
    queryFn: () => listScheduleExecutions(id!),
    enabled: !!id,
    refetchInterval: 15_000,
  });
}

/** Create a new build schedule. */
export function useCreateSchedule() {
  const queryClient = useQueryClient();

  return useMutation<BuildSchedule, Error, ScheduleCreateRequest>({
    mutationFn: createSchedule,
    onSuccess: (newSchedule) => {
      queryClient.invalidateQueries({ queryKey: scheduleKeys.all });
      queryClient.setQueryData(
        scheduleKeys.detail(newSchedule.id),
        newSchedule,
      );
    },
  });
}

/** Update an existing build schedule. */
export function useUpdateSchedule() {
  const queryClient = useQueryClient();

  return useMutation<
    BuildSchedule,
    Error,
    { id: string; data: ScheduleUpdateRequest }
  >({
    mutationFn: ({ id, data }) => updateSchedule(id, data),
    onSuccess: (updatedSchedule, { id }) => {
      queryClient.invalidateQueries({ queryKey: scheduleKeys.all });
      queryClient.setQueryData(
        scheduleKeys.detail(id),
        updatedSchedule,
      );
    },
  });
}

/** Delete a build schedule. */
export function useDeleteSchedule() {
  const queryClient = useQueryClient();

  return useMutation<void, Error, string>({
    mutationFn: deleteSchedule,
    onSuccess: (_data, id) => {
      queryClient.invalidateQueries({ queryKey: scheduleKeys.all });
      queryClient.removeQueries({ queryKey: scheduleKeys.detail(id) });
    },
  });
}

/** Pause an active schedule. */
export function usePauseSchedule() {
  const queryClient = useQueryClient();

  return useMutation<BuildSchedule, Error, string>({
    mutationFn: pauseSchedule,
    onSuccess: (updatedSchedule, id) => {
      queryClient.invalidateQueries({ queryKey: scheduleKeys.all });
      queryClient.setQueryData(
        scheduleKeys.detail(id),
        updatedSchedule,
      );
    },
  });
}

/** Resume a paused schedule. */
export function useResumeSchedule() {
  const queryClient = useQueryClient();

  return useMutation<BuildSchedule, Error, string>({
    mutationFn: resumeSchedule,
    onSuccess: (updatedSchedule, id) => {
      queryClient.invalidateQueries({ queryKey: scheduleKeys.all });
      queryClient.setQueryData(
        scheduleKeys.detail(id),
        updatedSchedule,
      );
    },
  });
}

/** Trigger an immediate execution of a schedule. */
export function useRunNow() {
  const queryClient = useQueryClient();

  return useMutation<ScheduleExecution, Error, string>({
    mutationFn: runScheduleNow,
    onSuccess: (_execution, id) => {
      queryClient.invalidateQueries({ queryKey: scheduleKeys.all });
      queryClient.invalidateQueries({
        queryKey: scheduleKeys.executions(id),
      });
      queryClient.invalidateQueries({
        queryKey: scheduleKeys.detail(id),
      });
    },
  });
}

File: backend/app/api/v1/schedules.py
"""Scheduled builds API endpoints.

Provides CRUD operations for build schedules plus lifecycle actions
(pause, resume, run-now) and execution history queries.
"""

import logging
import uuid
from datetime import datetime, timezone

from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy import func, select
from sqlalchemy.ext.asyncio import AsyncSession

from app.db.session import get_db
from app.models.schedule import BuildSchedule, ScheduleExecution
from app.schemas.schedule import (
    ExecutionListResponse,
    ExecutionResponse,
    ScheduleCreate,
    ScheduleDetailResponse,
    ScheduleListResponse,
    ScheduleResponse,
    ScheduleUpdate,
)
from app.services.cron_parser import get_next_run, validate_cron

logger = logging.getLogger(__name__)
router = APIRouter()


# ---------------------------------------------------------------------------
# Helpers
# ---------------------------------------------------------------------------


def _compute_next_run(schedule: BuildSchedule) -> datetime | None:
    """Compute the next_run_at value for a schedule.

    Returns ``None`` when the schedule should not run again (e.g.,
    one-time schedule already executed, or max_runs reached).
    """
    now = datetime.now(timezone.utc)

    if schedule.schedule_type == "one_time":
        if schedule.run_count > 0:
            return None
        return schedule.scheduled_at

    # Recurring
    if schedule.max_runs is not None and schedule.run_count >= schedule.max_runs:
        return None

    if schedule.cron_expression is None:
        return None

    reference = schedule.last_run_at or now
    try:
        return get_next_run(
            schedule.cron_expression,
            after=reference,
            timezone=schedule.timezone or "UTC",
        )
    except (ValueError, KeyError):
        logger.warning(
            "Failed to compute next_run_at for schedule %s with cron=%r tz=%r",
            schedule.id,
            schedule.cron_expression,
            schedule.timezone,
        )
        return None


# ---------------------------------------------------------------------------
# CRUD Endpoints
# ---------------------------------------------------------------------------


@router.post("/", response_model=ScheduleResponse, status_code=201)
async def create_schedule(
    request: ScheduleCreate,
    db: AsyncSession = Depends(get_db),
):
    """Create a new build schedule.

    For ``one_time`` schedules, ``scheduled_at`` is required.
    For ``recurring`` schedules, ``cron_expression`` is required and
    must be a valid 5-field cron string.
    """
    # Validate cron expression if recurring
    if request.schedule_type == "recurring" and request.cron_expression:
        if not validate_cron(request.cron_expression):
            raise HTTPException(
                status_code=422,
                detail=f"Invalid cron expression: {request.cron_expression!r}",
            )

    # Validate scheduled_at is in the future for one_time
    if request.schedule_type == "one_time" and request.scheduled_at:
        now = datetime.now(timezone.utc)
        scheduled_utc = request.scheduled_at
        if scheduled_utc.tzinfo is None:
            scheduled_utc = scheduled_utc.replace(tzinfo=timezone.utc)
        if scheduled_utc <= now:
            raise HTTPException(
                status_code=422,
                detail="scheduled_at must be in the future",
            )

    schedule = BuildSchedule(
        id=uuid.uuid4(),
        name=request.name,
        description=request.description,
        status="active",
        schedule_type=request.schedule_type,
        scheduled_at=request.scheduled_at,
        cron_expression=request.cron_expression,
        timezone=request.timezone,
        server_config=request.server_config,
        max_runs=request.max_runs,
        run_count=0,
        created_by=None,  # TODO: populate from auth context
    )

    # Compute initial next_run_at
    schedule.next_run_at = _compute_next_run(schedule)

    db.add(schedule)
    await db.flush()
    await db.refresh(schedule)

    logger.info(
        "Created build schedule: %s (%s) type=%s next_run=%s",
        schedule.name,
        schedule.id,
        schedule.schedule_type,
        schedule.next_run_at,
    )
    return schedule


@router.get("/", response_model=ScheduleListResponse)
async def list_schedules(
    status: str | None = Query(None, description="Filter by status"),
    schedule_type: str | None = Query(None, description="Filter by type"),
    skip: int = Query(0, ge=0),
    limit: int = Query(50, ge=1, le=200),
    db: AsyncSession = Depends(get_db),
):
    """List all build schedules with optional filtering."""
    query = select(BuildSchedule)
    count_query = select(func.count(BuildSchedule.id))

    if status:
        query = query.where(BuildSchedule.status == status)
        count_query = count_query.where(BuildSchedule.status == status)
    if schedule_type:
        query = query.where(BuildSchedule.schedule_type == schedule_type)
        count_query = count_query.where(BuildSchedule.schedule_type == schedule_type)

    query = query.order_by(BuildSchedule.created_at.desc()).offset(skip).limit(limit)

    result = await db.execute(query)
    schedules = result.scalars().all()

    total_result = await db.execute(count_query)
    total = total_result.scalar() or 0

    return ScheduleListResponse(schedules=schedules, total=total)


@router.get("/{schedule_id}", response_model=ScheduleDetailResponse)
async def get_schedule(
    schedule_id: uuid.UUID,
    db: AsyncSession = Depends(get_db),
):
    """Get a specific build schedule with its recent execution history."""
    result = await db.execute(
        select(BuildSchedule).where(BuildSchedule.id == schedule_id)
    )
    schedule = result.scalar_one_or_none()

    if schedule is None:
        raise HTTPException(status_code=404, detail="Build schedule not found")

    return schedule


@router.put("/{schedule_id}", response_model=ScheduleResponse)
async def update_schedule(
    schedule_id: uuid.UUID,
    request: ScheduleUpdate,
    db: AsyncSession = Depends(get_db),
):
    """Update an existing build schedule.

    Only ``active`` or ``paused`` schedules can be updated.  Completed
    and cancelled schedules are immutable.
    """
    result = await db.execute(
        select(BuildSchedule).where(BuildSchedule.id == schedule_id)
    )
    schedule = result.scalar_one_or_none()

    if schedule is None:
        raise HTTPException(status_code=404, detail="Build schedule not found")

    if schedule.status in ("completed", "cancelled"):
        raise HTTPException(
            status_code=400,
            detail=f"Cannot update a {schedule.status} schedule",
        )

    # Apply partial updates
    if request.name is not None:
        schedule.name = request.name
    if request.description is not None:
        schedule.description = request.description
    if request.scheduled_at is not None:
        schedule.scheduled_at = request.scheduled_at
    if request.cron_expression is not None:
        if not validate_cron(request.cron_expression):
            raise HTTPException(
                status_code=422,
                detail=f"Invalid cron expression: {request.cron_expression!r}",
            )
        schedule.cron_expression = request.cron_expression
    if request.timezone is not None:
        schedule.timezone = request.timezone
    if request.server_config is not None:
        schedule.server_config = request.server_config
    if request.max_runs is not None:
        schedule.max_runs = request.max_runs

    # Recompute next_run_at after any timing-related changes
    schedule.next_run_at = _compute_next_run(schedule)
    schedule.updated_at = datetime.now(timezone.utc)

    await db.flush()
    await db.refresh(schedule)

    logger.info("Updated build schedule: %s (%s)", schedule.name, schedule.id)
    return schedule


@router.delete("/{schedule_id}", response_model=ScheduleResponse)
async def delete_schedule(
    schedule_id: uuid.UUID,
    db: AsyncSession = Depends(get_db),
):
    """Cancel and soft-delete a build schedule.

    Sets the schedule status to ``cancelled`` and clears
    ``next_run_at`` so the periodic task will no longer pick it up.
    The record is preserved for audit purposes.
    """
    result = await db.execute(
        select(BuildSchedule).where(BuildSchedule.id == schedule_id)
    )
    schedule = result.scalar_one_or_none()

    if schedule is None:
        raise HTTPException(status_code=404, detail="Build schedule not found")

    if schedule.status == "cancelled":
        raise HTTPException(
            status_code=400,
            detail="Schedule is already cancelled",
        )

    schedule.status = "cancelled"
    schedule.next_run_at = None
    schedule.updated_at = datetime.now(timezone.utc)

    await db.flush()
    await db.refresh(schedule)

    logger.info("Cancelled build schedule: %s (%s)", schedule.name, schedule.id)
    return schedule


# ---------------------------------------------------------------------------
# Lifecycle Actions
# ---------------------------------------------------------------------------


@router.post("/{schedule_id}/pause", response_model=ScheduleResponse)
async def pause_schedule(
    schedule_id: uuid.UUID,
    db: AsyncSession = Depends(get_db),
):
    """Pause an active schedule.

    The schedule retains its configuration but ``next_run_at`` is
    cleared so the periodic task will skip it until it is resumed.
    """
    result = await db.execute(
        select(BuildSchedule).where(BuildSchedule.id == schedule_id)
    )
    schedule = result.scalar_one_or_none()

    if schedule is None:
        raise HTTPException(status_code=404, detail="Build schedule not found")

    if schedule.status != "active":
        raise HTTPException(
            status_code=400,
            detail=f"Only active schedules can be paused (current: {schedule.status})",
        )

    schedule.status = "paused"
    schedule.next_run_at = None
    schedule.updated_at = datetime.now(timezone.utc)

    await db.flush()
    await db.refresh(schedule)

    logger.info("Paused build schedule: %s (%s)", schedule.name, schedule.id)
    return schedule


@router.post("/{schedule_id}/resume", response_model=ScheduleResponse)
async def resume_schedule(
    schedule_id: uuid.UUID,
    db: AsyncSession = Depends(get_db),
):
    """Resume a paused schedule.

    Re-computes ``next_run_at`` from the current time and restores
    the schedule to ``active`` status.
    """
    result = await db.execute(
        select(BuildSchedule).where(BuildSchedule.id == schedule_id)
    )
    schedule = result.scalar_one_or_none()

    if schedule is None:
        raise HTTPException(status_code=404, detail="Build schedule not found")

    if schedule.status != "paused":
        raise HTTPException(
            status_code=400,
            detail=f"Only paused schedules can be resumed (current: {schedule.status})",
        )

    schedule.status = "active"
    schedule.next_run_at = _compute_next_run(schedule)
    schedule.updated_at = datetime.now(timezone.utc)

    await db.flush()
    await db.refresh(schedule)

    logger.info("Resumed build schedule: %s (%s)", schedule.name, schedule.id)
    return schedule


@router.post("/{schedule_id}/run-now", response_model=ExecutionResponse, status_code=202)
async def run_schedule_now(
    schedule_id: uuid.UUID,
    db: AsyncSession = Depends(get_db),
):
    """Trigger an immediate execution of a schedule.

    Creates a new ScheduleExecution with ``scheduled_for`` set to now
    and dispatches the Celery task.  Does not affect the regular
    schedule cadence.
    """
    result = await db.execute(
        select(BuildSchedule).where(BuildSchedule.id == schedule_id)
    )
    schedule = result.scalar_one_or_none()

    if schedule is None:
        raise HTTPException(status_code=404, detail="Build schedule not found")

    if schedule.status in ("cancelled", "completed"):
        raise HTTPException(
            status_code=400,
            detail=f"Cannot run a {schedule.status} schedule",
        )

    now = datetime.now(timezone.utc)

    execution = ScheduleExecution(
        id=uuid.uuid4(),
        schedule_id=schedule.id,
        status="pending",
        scheduled_for=now,
    )
    db.add(execution)
    await db.flush()
    await db.refresh(execution)

    # Dispatch Celery task
    try:
        from app.tasks.scheduler import execute_scheduled_build
        execute_scheduled_build.delay(str(schedule.id), str(execution.id))
    except ImportError:
        logger.warning("Celery scheduler tasks not yet available, skipping dispatch")

    logger.info(
        "Triggered immediate execution for schedule %s (%s), execution %s",
        schedule.name,
        schedule.id,
        execution.id,
    )
    return execution


# ---------------------------------------------------------------------------
# Execution History
# ---------------------------------------------------------------------------


@router.get("/{schedule_id}/executions", response_model=ExecutionListResponse)
async def list_executions(
    schedule_id: uuid.UUID,
    status: str | None = Query(None, description="Filter by execution status"),
    skip: int = Query(0, ge=0),
    limit: int = Query(50, ge=1, le=200),
    db: AsyncSession = Depends(get_db),
):
    """List execution history for a specific schedule."""
    # Verify the schedule exists
    sched_result = await db.execute(
        select(BuildSchedule.id).where(BuildSchedule.id == schedule_id)
    )
    if sched_result.scalar_one_or_none() is None:
        raise HTTPException(status_code=404, detail="Build schedule not found")

    query = select(ScheduleExecution).where(
        ScheduleExecution.schedule_id == schedule_id
    )
    count_query = select(func.count(ScheduleExecution.id)).where(
        ScheduleExecution.schedule_id == schedule_id
    )

    if status:
        query = query.where(ScheduleExecution.status == status)
        count_query = count_query.where(ScheduleExecution.status == status)

    query = (
        query.order_by(ScheduleExecution.scheduled_for.desc())
        .offset(skip)
        .limit(limit)
    )

    result = await db.execute(query)
    executions = result.scalars().all()

    total_result = await db.execute(count_query)
    total = total_result.scalar() or 0

    return ExecutionListResponse(executions=executions, total=total)
